sympy/utilities/iterables.py

```diff
@@ -2285,14 +2285,244 @@ def generate_involutions(n):
             yield p
 
 
+def multiset_derangements(s):
+    """Generate derangements of the elements of s *in place*.
+
+    Examples
+    ========
+
+    >>> from sympy.utilities.iterables import multiset_derangements, uniq
+
+    Because the derangements of multisets (not sets) are generated
+    in place, copies of the return value must be made if a collection
+    of derangements is desired or else all values will be the same:
+
+    >>> list(uniq([i for i in multiset_derangements('1233')]))
+    [['3', '3', '2', '1']]
+    >>> [i.copy() for i in multiset_derangements('1233')]
+    [['3', '3', '1', '2'], ['3', '3', '2', '1']]
+    """
+    ms = multiset(s)
+    mx = max(ms.values())
+    n = len(s)
+    # impossible case
+    if mx*2 > n:
+        return
+    rv = [None]*n
+    # special cases
+
+    # 1) singletons
+    if len(ms) == n:
+        for p in generate_derangements(s):
+            yield p
+        return
+
+    # 2) aaabbb-like
```

Comment:
I don't think case #2 is needed any more -- the finish_derangements code handles these cases efficiently.
For example:
```
assert(['b', 'a'] * 25 ==
           list(list(d) for d in
                multiset_derangements_mod2(['a', 'b'] * 25)) [0]
```
returns instantly.